lib/active_merchant/billing/gateways/nuvei.rb

```diff
@@ -0,0 +1,137 @@
+module ActiveMerchant
+  module Billing
+    class NuveiGateway < Gateway
+      self.test_url = 'https://ppp-test.nuvei.com/ppp/api/v1'
+      self.live_url = 'https://secure.safecharge.com/ppp/api/v1'
+
+      self.supported_countries = %w[US CA IN NZ GB AU US]
+      self.default_currency = 'USD'
+      self.money_format = :cents
+      self.supported_cardtypes = %i[visa master american_express discover union_pay]
+      self.currencies_without_fractions = %w[CLP KRW JPY ISK MMK PYG UGX VND XAF XOF]
+      self.homepage_url = 'https://www.nuvei.com/'
+      self.display_name = 'Nuvei'
+
+      ENDPOINTS_MAPPING = {
+        authenticate: '/getSessionToken',
+        purchase: '/payment', # /authorize with transactionType: "Auth"
+        capture: '/settleTransaction',
+        refund: '/refundTransaction',
+        void: '/voidTransaction',
+        general_credit: '/payout'
+      }
+
+      def initialize(options = {})
+        requires!(options, :merchant_id, :merchant_site_id, :secret_key)
+        super
+      end
+
+      def authorize(money, payment, options = {}); end
+
+      def purchase(money, payment, options = {}); end
+
+      def capture(money, authorization, options = {}); end
+
+      def refund(money, authorization, options = {}); end
+
+      def void(authorization, options = {}); end
+
+      def credit(money, payment, options = {}); end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((Authorization: Bearer )[a-zA-Z0-9._-]+)i, '\1[FILTERED]').
+          gsub(%r((\"cardNumber\\\":\\\")\d+), '\1[FILTERED]').
+          gsub(%r((\"cardCvv\\\":\\\")\d+), '\1[FILTERED]').
+          gsub(%r((\"cardHolderName\\\":\\\")\w+), '\1[FILTERED]')
+      end
+
+      private
+
+      def add_payment(post, payment); end
+
+      def commit(action, post, authorization = nil, method = :post)
+        MultiResponse.run do |r|
+          r.process { fetch_session_token } unless session_token_valid?
+          r.process do
+            api_request(action, post, authorization, method).tap do |response|
+              response.params.merge!(@options.slice(:session_token, :token_expires)) if @options[:new_credentials]
+            end
+          end
+        end
+      end
+
+      def session_token_valid?
+        @options[:session_token] && @options[:token_expires] && Time.now <= @options[:token_expires]
+      end
+
+      def fetch_session_token
```

Comment:
💬 Comment:

Unlike FlexCharge this method needs to do more work to build the request, maybe a `build_xxxxx` method would be nice to separte concerns and also to add proper unit testing for the code that builds the request body.

Reply:
Good suggestion, I will take a look on it, thanks